### PR TITLE
External module bundling

### PR DESCRIPTION
**Pull request by weswigham**
Implements #4434 loosely based on the feedback in the issue and some internal feedback.

Changes from the original proposal:
* The internal loader is commonjs based
* The internal loader intercepts all imports, attempts to resolve them within the bundle, then passes them off to the external loader (if available)

Things to do (aside from integrate feedback):
- [x] many many tests
- [ ] consider es6 module compatible emit
- [x] improve systemjs support - we don't support full es6 semantics even when targeting systemjs with this loader, and we can't handle dynamic systemjs external dependencies (since they're async)
- [ ] other internal loader options?

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -167,6 +245,64 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
             /** Called once the emit of the node is done */
             let emitEnd = function (node: Node) { };
 
+            /** Called to build the opening part of a bundled module wrapper */
+            let emitBundleWrapperStart: Map<Function>  = {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: This should be `Map<() => void>`.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -167,6 +245,64 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
             /** Called once the emit of the node is done */
             let emitEnd = function (node: Node) { };
 
+            /** Called to build the opening part of a bundled module wrapper */
+            let emitBundleWrapperStart: Map<Function>  = {
+                [ModuleKind.CommonJS]: () => {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Use method syntax here:

```TypeScript
[ModuleKind.CommonJS]() {
    write("module.exports = ");
}
```

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -198,11 +334,32 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
             }
             else {
+                let originalKind = compilerOptions.module;
+                if (compilerOptions.bundle) {
+                    compilerOptions.module = ModuleKind.CommonJS;
+                    emitBundleWrapperStart[originalKind]();
+                    write("(function() {");
+                    increaseIndent();
+                    writeLines(defineHelper);
+                    forEach(host.getSourceFiles(), file => {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Turn this into a `for-of`

Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `x => f(x)` can be converted into `f`

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -198,11 +334,32 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                        emitEmitHelpers(file);
+                    });
+                    writeLine();
+                }
                 forEach(host.getSourceFiles(), sourceFile => {
                     if (!isExternalModuleOrDeclarationFile(sourceFile)) {
                         emitSourceFile(sourceFile);
+                    } else if (compilerOptions.bundle && isExternalModule(sourceFile)) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `else` on the next line

File: src/compiler/emitter.ts [hunk 5]
```diff
@@ -198,11 +334,32 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                        emitBundledFile(sourceFile);
                     }
                 });
+                if (compilerOptions.bundle) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Newine above

File: src/compiler/emitter.ts [hunk 6]
```diff
@@ -6314,6 +6461,37 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+
+                writeLine();
+
+                // define an export star helper function
+                write(`function ${exportStarFunction}(m) {`);
+                increaseIndent();
+                writeLine();
+                write(`var exports = {};`);
+                writeLine();
+                write(`for(var n in m) {`);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `for (var n in m) {`

File: src/compiler/emitter.ts [hunk 7]
```diff
@@ -167,6 +245,64 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                    write("return ");
+                },
+                [ModuleKind.System]: () => { //TODO: handle external deps with system
+                    currentSourceFile = <any>{identifiers: []};
+                    exportFunctionForFile = makeUniqueName("exports");
+                    writeLine();
+                    write(`System.register([], function(${exportFunctionForFile}) {`);
+                    writeLine();
+                    increaseIndent();
+                    let exportStarName = emitExportStarFunction(undefined);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: `/* localNames */`

File: src/compiler/emitter.ts [hunk 8]
```diff
@@ -167,6 +245,64 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+                },
+                [ModuleKind.UMD]: () => {
+                    writeLines(umdHelper);
+                    write("[\"require\"], function(require){");
+                    writeLine();
+                    increaseIndent();
+                    write("return ");
+                },
+                [ModuleKind.System]: () => { //TODO: handle external deps with system
+                    currentSourceFile = <any>{identifiers: []};
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I don't know what you're trying to achieve here

File: src/compiler/emitter.ts [hunk 9]
```diff
@@ -214,6 +371,26 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                 exportFunctionForFile = undefined;
                 emit(sourceFile);
             }
+            
+            function emitBundledFile(sourceFile: SourceFile): void {
+                let canonicalName = removeFileExtension(
+                    getRelativePathToDirectoryOrUrl(
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Extract this out to another local variable.

File: src/compiler/emitter.ts (hunk 6 above)
Comment (weswigham) [inline src/compiler/emitter.ts]: I actually just moved this from an inner scope so I could reuse it - changing this is going to change some baselines. :3

File: src/compiler/emitter.ts [hunk 10]
```diff
@@ -214,6 +371,26 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
                 emit(sourceFile);
             }
+            
+            function emitBundledFile(sourceFile: SourceFile): void {
+                let canonicalName = removeFileExtension(
+                    getRelativePathToDirectoryOrUrl(
+                        host.getCurrentDirectory(),
+                        sourceFile.fileName,
+                        host.getCurrentDirectory(),
+                        host.getCanonicalFileName,
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: You actually need to enclose this in an arrow function to ensure that you maintain its `this` context.

File: src/compiler/emitter.ts [hunk 11]
```diff
@@ -71,12 +147,14 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
 
         if (targetSourceFile === undefined) {
-            forEach(host.getSourceFiles(), sourceFile => {
-                if (shouldEmitToOwnFile(sourceFile, compilerOptions)) {
-                    let jsFilePath = getOwnEmitOutputFilePath(sourceFile, host, shouldEmitJsx(sourceFile) ? ".jsx" : ".js");
-                    emitFile(jsFilePath, sourceFile);
-                }
-            });
+            if (!compilerOptions.bundle) {
+                forEach(host.getSourceFiles(), sourceFile => {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: I'd use a for-of here

File: src/compiler/emitter.ts [hunk 12]
```diff
@@ -62,6 +62,82 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
+        }
+    },
+    resolvePath = function(base, name) {
+        var result = normalizeSlashes(base).split("/");
+        result.pop();
+        forEach(normalizeSlashes(name).split("/"), function(part, index) {
+            switch(part) {
+                case "":
+                case ".":
+                break;
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Indent from case clauses

**Comment by gulbanana** [comment]
does this implementation support the use case of transitive multi-file bundles?

scenario: 
typescript project A contains Class1 and Class2. Class2 depends upon NpmModule
typescript project B contains Class3, depending on Class1 or Class2.

the structure we really want is a namespace for each of A and B and something 'dll-like' which wraps them up. however, Class2 must be defined in an external module to be able to import NpmModule. this causes a cascade of module usage:
Class1 must be an external module so that it can be bundled with Class2 into bundle A. Class3 must now be an external module so that it can import A, and from A, Class1 or Class2.

in reality project A and B might each have dozens or hundreds of classes, none of which have any reason to be defined in modules and internally microloaded other than the fact that some leaf of the internal dependency tree uses NpmModule. is this going to work at all?

**Comment by weswigham** [comment]
In the end, this at present does support offloading module dependencies to the external module loader when in-bundle resolution fails (the bundle consisting of every TS file in your project, rather than just those deemed 'reachable' from the entrypoint). This passthru works well for `commonjs` and even okayish for `amd`, but for systemjs it doesn't work particularly well - in fact, it doesn't deal with external/non-ts dependencies well at all.

On that note - should our bundler bundle in reachable non-ts dependencies, a la browserify or webpack? Or is the TS project the exclusive scope of the bundle, and external bits are left to the user to package/bundle? A practical example would be bundling angular - I've been able to bundle it with the bundler in #4754 without it bundling in `Rx`, `zone.js`, or `reflect-metadata` and it builds... but there is a call to `require('rx')` that at present goes unfulfilled unless the loader using the bundle provides it. (And `reflect-metadata` needs to get included in the page, since it patches the global object.)

In either case, in situations like `systemjs` or `amd` style dependency emit, it would be useful to be able to specify external dependencies in a hash of internal name to external name (this isn't useful for `commonjs` emit since we can just yield to the platform `require`):
```js
"bundleDependencies": {
  "angular2/angular2": "angular2",
  "jQuery": "jquery"
}
```

to allow your bundle-external import statements
```ts
import * as ng from 'angular2/angular2';
import * as $ from 'jQuery';
import SomeComponent from './somecomponent';
ng.bootstrap(SomeComponent);
$('title').html('SomeComponent Test');
```
to be depended upon correctly in the resulting bundle:
```js
System.register(["angular2", "jquery"], function(exports_1) {
  var angular2_min_js;
  var jquery_min_js;
  //...
  return {
    setters: [function(m){angular2_min_js = m;}, function(m){jquery_min_js = m;}],
    execute: function() {
      //...
    }
  };
});
```

I propose adding this option hash, rather than traversing imports and attempting to identify what is bundle local ourselves, to support the same dynamic cases as mentioned in the original issue, but with bundle-external modules.

Granted, we could do both - attempt automagic external dependency finding and defer to such a hash option if present.

**Comment by weswigham** [comment]
After some conversation, I think we've decided to simply autodetect external imports - I'm considering confirmed external imports as external and `undefined` imports (imports TS couldn't find) external.

It also looks like we're going to just bundle TS, and only TS within our own package - so I'll need to test that we don't bundle anything across package bounds.

**Comment by DanielRosenwasser** [comment]
Can you include work as part of #1544 in this change so that we issue an error and quit when

1. `--out` or `--outFile` are used and
2. Either
    1. `--module` is specified
    2. You are compiling in ES6 mode and any of your files has an external module indicator.

Or some variation of the above.

**Comment by weswigham** [comment]
@DanielRosenwasser You sure you didn't want to tack that on to #4811 instead?

**Pull request closed without merging**